### ISSUE TITLE
Integration seems broken under HA 2024.9.2

### ISSUE THREAD

**Issue by gmint** [Schnitzel/hass-miner#374]
Under HA 2024.9.2 I am able to add the integration and add miners (two S19s running BOS 24.0.3.1) but there are no entities available. I had this problem in a previous version and I was going to report it but removing and re-adding the integration and the devices resolved the issue. However, as of 2024.9.2 (running on VirtualBox) this no longer works and the integration seems to be non-functional.

**Comment by b-rowan** [comment]
This is on `v1.1.14`?  Anything in the logs?

**Comment by gmint** [comment]
Sorry, yes, I tried under 1.1.14 and 1.1.14rc1. Here's what I've got from the logs:

Logger: homeassistant.helpers.entity_platform
Source: helpers/entity_platform.py:415
First occurred: 11:20:41 AM (15 occurrences)
Last logged: 11:38:19 AM

miner raises exception ConfigEntryNotReady in forwarded platform sensor; Instead raise ConfigEntryNotReady before calling async_forward_entry_setups
miner raises exception ConfigEntryNotReady in forwarded platform switch; Instead raise ConfigEntryNotReady before calling async_forward_entry_setups
miner raises exception ConfigEntryNotReady in forwarded platform number; Instead raise ConfigEntryNotReady before calling async_forward_entry_setups

**Comment by b-rowan** [comment]
Alright, config entry timing again.  Ill see what I can do about this, but this specific issue is pretty annoying since it doesnt seem to be well documented where im supposed to raise that error when the miner doesn't respond...

**Comment by gmint** [comment]
Is there anything I can do to workaround? I'm having this issue on one HA install, but it doesn't seem to be happening on my other install?

**Comment by b-rowan** [comment]
> Is there anything I can do to workaround? I'm having this issue on one HA install, but it doesn't seem to be happening on my other install. Just trying to make your life a little easier if possible.

Not really, its likely this will happen on multiple versions, and im not sure the exact circumstances that cause it.  I'm looking into it, as others have raised this issue as well, just a bit difficult for me to emulate HASS without just spinning up an instance.

**Comment by gmint** [comment]
Understood. I feel your pain. I tried my hand at developing an integration and I ultimately quit because I got no help from the HA devs (and they were actually quite rude).

**Comment by b-rowan** [comment]
> Understood. I feel your pain. I tried my hand at developing an integration and I ultimately quit because I got no help from the HA devs (and they were actually quite rude).

Yeah, I've asked about a couple things in the discord and not gotten any response.  Quite disappointing, as HASS is a cool system.

For reference, here is the docs on how to do this "properly" - https://developers.home-assistant.io/docs/integration_setup_failures/#handling-offline-or-unavailable-devices-and-services

And the implementation in this repo - 
https://github.com/Schnitzel/hass-miner/blob/ba6a22063674ad2c062cbc0e9118ec137a9be119/custom_components/miner/__init__.py#L18-L34

From the docs - 
> Raise the ConfigEntryNotReady exception from async_setup_entry in the integration's `__init__.py`, and Home Assistant will automatically take care of retrying set up later.

Here, we have the `async_setup_entry` function in `__init__py` and it is raising `ConfigEntryNotReady`. 

Log reports -
> raise ConfigEntryNotReady before calling async_forward_entry_setups

This is irrelevant, as you can see above this is raised way before `async_forward_entry_setup`.

TL;DR: repo does EXACTLY as the documentation says, and it doesn't work?

**Comment by b-rowan** [comment]
From what I can see, it might be some weird conflict between `DataUpdateCoordinator` and the `__init__.py` `async_setup_entry` seems like they should be combined somehow?

https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

**Comment by b-rowan** [comment]
Fixed with a PR I think, I'm going to release a rc version for testing.

**Comment by b-rowan** [comment]
Try `v1.1.15rc1`

**Comment by gmint** [comment]
It looks to be retrying automatically now but I'm now getting: 

Failed Setup, will retry: Unknown error

This time I'm not really seeing anything in the logs.

**Comment by b-rowan** [comment]
> It looks to be retrying automatically now but I'm now getting: 
> 
> 
> 
> Failed Setup, will retry: Unknown error
> 
> 
> 
> This time I'm not really seeing anything in the logs.

Ok, that's a good sign, it's raising a valid error now, that's from the coordinator.  Something is happening that it can't identify...

**Comment by b-rowan** [comment]
Debugging fix coming for that

**Comment by gmint** [comment]
Let me know, happy to help however I can.

**Comment by b-rowan** [comment]
Try with `v1.1.15rc2`, debugging info should be improved.  You can also put the integration into debug mode.

**Comment by gmint** [comment]
Here's what I got when I removed and added back one of my miners:

2024-09-17 14:13:36.999 ERROR (MainThread) [homeassistant.components.mqtt.client] Failed to connect to MQTT server due to exception: [Errno -5] Name has no usable address
2024-09-17 14:13:53.323 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.152
2024-09-17 14:13:53.654 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer2 data in 0.356 seconds (success: False)
2024-09-17 14:13:53.725 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.151
2024-09-17 14:13:53.963 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer1 data in 0.261 seconds (success: False)
2024-09-17 14:14:14.086 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.151
2024-09-17 14:14:14.224 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.152
2024-09-17 14:14:14.305 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer1 data in 0.242 seconds (success: False)
2024-09-17 14:14:14.443 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer2 data in 0.281 seconds (success: False)
2024-09-17 14:14:54.563 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.151
2024-09-17 14:14:54.623 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.152
2024-09-17 14:14:54.843 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer1 data in 0.322 seconds (success: False)
2024-09-17 14:14:54.845 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer2 data in 0.249 seconds (success: False)
2024-09-17 14:16:15.114 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.151
2024-09-17 14:16:15.298 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.152
2024-09-17 14:16:15.365 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer1 data in 0.277 seconds (success: False)
2024-09-17 14:16:15.528 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer2 data in 0.258 seconds (success: False)
2024-09-17 14:17:18.222 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.151
2024-09-17 14:17:18.443 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer1 data in 0.249 seconds (success: False)
2024-09-17 14:17:23.863 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.151
2024-09-17 14:17:24.096 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer1 data in 0.263 seconds (success: False)

**Comment by b-rowan** [comment]
Bumped to RC3 with explicit `logging.exception` call...

**Comment by gmint** [comment]
I think this will help you:

File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 1717, in to_pydict
value = getattr(self, field_name)
^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 840, in __getattribute__
raise AttributeError(
AttributeError: '_id' is set to None, not 'id'
2024-09-17 14:31:03.352 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer1 data in 0.279 seconds (success: False)
2024-09-17 14:31:03.558 ERROR (MainThread) [custom_components.miner.coordinator] '_id' is set to None, not 'id'
Traceback (most recent call last):
File "/config/custom_components/miner/coordinator.py", line 85, in _async_update_data
miner_data = await self.miner.get_data(
^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/miners/base.py", line 535, in get_data
gathered_data = await self._get_data(
^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/miners/base.py", line 460, in _get_data
await asyncio.gather(rpc_command_task, web_command_task)
File "/usr/local/lib/python3.12/site-packages/pyasic/web/braiins_os/boser.py", line 88, in multicommand
result[cmd] = await tasks[cmd]
^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/web/braiins_os/boser.py", line 202, in get_cooling_state
return await self.send_command(
^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/web/braiins_os/boser.py", line 114, in send_command
return (await endpoint(message, metadata=metadata)).to_pydict()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 1756, in to_pydict
output[cased_name] = value.to_pydict(casing, include_default_values)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 1717, in to_pydict
value = getattr(self, field_name)
^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 840, in __getattribute__
raise AttributeError(
AttributeError: '_id' is set to None, not 'id'
2024-09-17 14:31:03.560 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer2 data in 0.285 seconds (success: False)
2024-09-17 14:32:23.621 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.151
2024-09-17 14:32:23.836 ERROR (MainThread) [custom_components.miner.coordinator] '_id' is set to None, not 'id'
Traceback (most recent call last):
File "/config/custom_components/miner/coordinator.py", line 85, in _async_update_data
miner_data = await self.miner.get_data(
^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/miners/base.py", line 535, in get_data
gathered_data = await self._get_data(
^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/miners/base.py", line 460, in _get_data
await asyncio.gather(rpc_command_task, web_command_task)
File "/usr/local/lib/python3.12/site-packages/pyasic/web/braiins_os/boser.py", line 88, in multicommand
result[cmd] = await tasks[cmd]
^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/web/braiins_os/boser.py", line 202, in get_cooling_state
return await self.send_command(
^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/web/braiins_os/boser.py", line 114, in send_command
return (await endpoint(message, metadata=metadata)).to_pydict()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 1756, in to_pydict
output[cased_name] = value.to_pydict(casing, include_default_values)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 1717, in to_pydict
value = getattr(self, field_name)
^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 840, in __getattribute__
raise AttributeError(
AttributeError: '_id' is set to None, not 'id'
2024-09-17 14:32:23.838 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer1 data in 0.243 seconds (success: False)
2024-09-17 14:32:24.022 DEBUG (MainThread) [custom_components.miner.coordinator] Found miner: S19 (BOS+): 10.0.1.152
2024-09-17 14:32:24.247 ERROR (MainThread) [custom_components.miner.coordinator] '_id' is set to None, not 'id'
Traceback (most recent call last):
File "/config/custom_components/miner/coordinator.py", line 85, in _async_update_data
miner_data = await self.miner.get_data(
^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/miners/base.py", line 535, in get_data
gathered_data = await self._get_data(
^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/miners/base.py", line 460, in _get_data
await asyncio.gather(rpc_command_task, web_command_task)
File "/usr/local/lib/python3.12/site-packages/pyasic/web/braiins_os/boser.py", line 88, in multicommand
result[cmd] = await tasks[cmd]
^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/web/braiins_os/boser.py", line 202, in get_cooling_state
return await self.send_command(
^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/pyasic/web/braiins_os/boser.py", line 114, in send_command
return (await endpoint(message, metadata=metadata)).to_pydict()
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 1756, in to_pydict
output[cased_name] = value.to_pydict(casing, include_default_values)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 1717, in to_pydict
value = getattr(self, field_name)
^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.12/site-packages/betterproto/__init__.py", line 840, in __getattribute__
raise AttributeError(
AttributeError: '_id' is set to None, not 'id'
2024-09-17 14:32:24.248 DEBUG (MainThread) [custom_components.miner.coordinator] Finished fetching antminer2 data in 0.257 seconds (success: False)

**Comment by b-rowan** [comment]
Betterproto.  I hate GRPC lol.

This will have to be fixed in pyasic, Ill see what I can do.

**Comment by gmint** [comment]
That is something I'm not familiar with personally. But thanks for looking into this!

**Comment by b-rowan** [comment]
> That is something I'm not familiar with personally. But thanks for looking into this!

GRPC is the communication protocol used by recent version of BOS+ for their API.  Not sure where that error is coming from, but its being raised here in betterproto - 

https://github.com/danielgtaylor/python-betterproto/blob/8d25c96cea597b9555e1ba3736b0b287cd3172ce/src/betterproto/__init__.py#L840-L844

I may just need to re-compile the proto files for BOS.

**Comment by b-rowan** [comment]
Can you try updating your miners actually?  Possible the compiled files in pyasic are TOO new for your miner...

**Comment by gmint** [comment]
I actually did that in the process of trying to fix this, I'm now on 24.08.1-plus

**Comment by b-rowan** [comment]
> I actually did that in the process of trying to fix this, I'm now on 24.08.1-plus

Ok, ill have to re-compile them then I suppose.

**Comment by gmint** [comment]
BTW, I should mention that your integration is really awesome. I use it to turn on my bitcoin miners only when my solar panels are making enough energy to not take from the grid so mining is still profitable.

**Comment by b-rowan** [comment]
> BTW, I should mention that your integration is really awesome. I use it to turn on my bitcoin miners only when my solar panels are making enough energy to not take from the grid so mining is still profitable.

Thanks, although not entirely my integration, as I just maintain it.  Was a great idea on the part of @Schnitzel.

**Comment by gmint** [comment]
Either way, it's really useful. I also use it to control my bitcoin miner that heats my hot tub...

**Comment by b-rowan** [comment]
> Either way, it's really useful. I also use it to control my bitcoin miner that heats my hot tub...

Thats pretty awesome.  I'm never sure how many people are doing projects like this, but we need to see more of this stuff in the industry.  The more people find use for bitcoin mining the more innovation we get in the home mining space.

**Comment by gmint** [comment]
I agree. Honestly, this project was great to find because I also use my miners to heat my business during the winter. Previously I was just doing it with a shell script and some TCP commands, but this is way better because I can use the status to turn on and off my external fans (the built in ones are too loud).

**Comment by b-rowan** [comment]
Not sure of a working version, I can look and see if I can figure out what version it was compiled for, but I don't think there are any indicators in the sources...

I would try 24.04 or 24.06 maybe?

**Comment by gmint** [comment]
Just FYI, I was on BOS 24.0.3.1 when I was initially experiencing this issue.

**Comment by b-rowan** [comment]
Ok, issue was with betterproto 2.0.0b7, I have pinned it in pyasic now.  Technically the BOS+ proto files aren't following proper GRPC protocol now, the `oneof` values should always have a value, but whatever.  Works with 2.0.0b6.

This should be fixed with v1.1.15rc4, let me know if that works and ill move it from pre to latest.

**Comment by gmint** [comment]
Ok, it's working on rc4! One small note which isn't particularly consequential but I did have to restart my miners and toggle the active switch once or twice to seemingly get things in sync and functioning. When I added the device, it showed as on in HA, but the device was not running and seemed to be in some sort of stuck state until I restarted. No idea if this is related to the integration or not, but I thought I would mention it.

**Comment by b-rowan** [comment]
Possibly a pyasic bug, but at least it's less concerning haha

**Comment by gmint** [comment]
I agree 100%

**Comment by b-rowan** [comment]
Ok, fixed in v1.1.15 on stable.  Thanks for all your help!
